### PR TITLE
perf/port optimizations from Lambdaworks

### PR DESCRIPTION
**Pull request by jotabulacios**
This PR ports the Goldilocks field arithmetic optimizations from Lambdaworks (compiler hints, branch prediction, and x86_64 inline assembly for the base field) and extends them with fused dot-product accumulation for extension field arithmetic.

#### Base field 
- **add/sub**: Uses `assume()` + `branch_hint()` for rare double overflow/underflow cases, avoiding `cmov` on predictable branches.  
- **reduce128**: Applies the same branch-hint pattern for rare borrow/carry scenarios.  
- **add_no_canonicalize_trashing_input**: Uses x86_64 `add` + `sbb reg32, reg32` for a 2-instruction modular addition, with a portable fallback.

#### Fused dot-product and Fp3 rewrite 

`dot_product_2` and `dot_product_3` accumulate multiple `u64 × u64` products as `u128` and perform a single reduction instead of reducing per product.

This enables a rewrite of **Fp3**

- **Fp3 multiplication**:  
  Karatsuba (`6 muls + 6 reduce128`) → schoolbook + `dot_product_3` (`9 muls + 3 reduce128`).  
  Extra multiplications are cheaper than the saved reductions.

- **Fp3 squaring**:  
  `6 → 3 reduce128` using `dot_product_2`.

#### Testing

- Edge-case tests for non-canonical inputs  
- Expanded fuzz target for raw `u64` coverage  
- Inline proofs for unsafe `assume` invariants

### CODE REVIEW

File: crypto/math/src/field/goldilocks.rs [hunk 1]
```diff
@@ -142,37 +191,137 @@ impl IsField for GoldilocksField {
...
+    let prod0 = (a0 as u128) * (b0 as u128);
+    let prod1 = (a1 as u128) * (b1 as u128);
+    let (sum, overflow) = prod0.overflowing_add(prod1);
+
+    let reduced = reduce128(sum);
+
+    if overflow {
+        // True value is sum + 2^128. Since 2^128 mod p = EPSILON^2,
+        // add EPSILON^2 = (2^32-1)^2 = 2^64 - 2^33 + 1.
+        // This value is < p, and reduced < 2p, so their sum < 3p < 2^64 + p,
```
[the remark this answers is not in the document]
Comment (jotabulacios) [reply crypto/math/src/field/goldilocks.rs]: Solved in [61ced83](https://github.com/yetanotherco/lambda_vm/pull/456/commits/61ced831f01a3929ff174f2dbde6412dd28c61fb)

File: crypto/math/src/field/goldilocks.rs [hunk 2]
```diff
@@ -142,37 +184,60 @@ impl IsField for GoldilocksField {
...
+            "add {0}, {1}",
+            // sbb {1:e}, {1:e} sets the low 32 bits to 0xFFFFFFFF on carry (= NEG_ORDER),
+            // or 0 otherwise. The high 32 bits are zeroed by the 32-bit register write.
+            "sbb {1:e}, {1:e}",
+            inlateout(reg) x => res_wrapped,
+            inlateout(reg) y => adjustment,
+            options(pure, nomem, nostack),
+        );
     }
+    res_wrapped + adjustment
```
[the remark this answers is not in the document]
Comment (jotabulacios) [reply crypto/math/src/field/goldilocks.rs:240]: Intentional : panics in debug on precondition violation, which is better than wrapping_add silently returning wrong results.

File: crypto/math/src/field/goldilocks.rs [hunk 3]
```diff
@@ -35,32 +72,37 @@ pub struct GoldilocksField;
...
-        let (diff, under2) = diff.overflowing_sub((under as u64) * EPSILON);
-        // Second underflow is rare but possible
-        if under2 {
-            diff.wrapping_sub(EPSILON)
-        } else {
-            diff
+        let (mut diff, under) = diff.overflowing_sub((under as u64) * EPSILON);
+        if under {
+            unsafe {
+                assume(*a < EPSILON - 1 && *b > GOLDILOCKS_PRIME);
```
[the remark this answers is not in the document]
Comment (jotabulacios) [reply crypto/math/src/field/goldilocks.rs:107]: Addressed in 787dfed.  added inline proof explaining why a < EPSILON - 1 is tight (at a = EPSILON - 1, diff1 = EPSILON exactly, so no second underflow)

File: crypto/math/src/field/goldilocks.rs [hunk 4]
```diff
@@ -35,32 +72,37 @@ pub struct GoldilocksField;
...
-        // Second overflow is rare but possible
-        if over2 {
-            sum.wrapping_add(EPSILON)
-        } else {
-            sum
+        let (mut sum, over) = sum.overflowing_add((over as u64) * EPSILON);
+        if over {
+            // Double overflow requires both inputs > ORDER, which is exceedingly rare.
+            unsafe {
+                assume(*a > GOLDILOCKS_PRIME && *b > GOLDILOCKS_PRIME);
```
[the remark this answers is not in the document]
Comment (jotabulacios) [reply crypto/math/src/field/goldilocks.rs:87]: Addressed in 787dfed. Added inline proof above the assume explaining the threshold.

**Review by diegokingston — APPROVED** [review]
